theano/sandbox/cuda/basic_ops.py

```diff
@@ -3340,6 +3319,75 @@ def c_code(self, node, name, inputs, out_, sub):
                 %(fail)s;
             }
         }
+        """ % locals()
+        return str
+
+    def infer_shape(self, node, input_shapes):
+        return [node.inputs]
+
+    def grad(self, inputs, grads):
+        gout, = grads
+        return [None for i in inputs]
+
+    def c_code_cache_version(self):
+        return (1,)
+
+    def do_constant_folding(self, node):
+        # XXX: anything needed here?
+        return False
+
+gpu_alloc_empty = GpuAllocEmpty()
+
+
+class GpuAlloc(GpuAllocEmpty):
+    """Implement Alloc on the gpu.
+
+    The memset_0 param is an optimization. When True, we call
+    cudaMemset that is faster.
+
+    """
+    def __init__(self, memset_0=False):
+        self.memset_0 = memset_0
+
+    def __eq__(self, other):
+        return type(self) == type(other) and self.memset_0 == other.memset_0
+
+    def __hash__(self):
+        return hash(type(self)) ^ hash(self.memset_0)
+
+    def __str__(self):
+        #Hide the memset parameter when not used to prevent confusion.
+        if self.memset_0:
+            s = "%s{memset_0=%s}" % (self.__class__.__name__, self.memset_0)
+        else:
+            s = self.__class__.__name__
+        return s
+
+    def make_node(self, value, *shape):
+        node = super(GpuAlloc, self).make_node(*shape)
+        node.outputs[0].owner = None
```

Comment:
It's not just the output type, though, it's also the shape. I first had a helper method `get_apply_args()` like this:
```python
class GpuAllocEmpty(GpuOp):
    def get_apply_args(self, *shape):
        ...
        return inputs, outputs
    def make_node(self, *args):
        return Apply(self.get_apply_args(*args))
class GpuAlloc(GpuAllocEmpty):
    def get_apply_args(self, value, *shape):
        inputs, outputs = super(GpuAlloc, self).get_apply_args(*shape)
        v = as_cuda_ndarray_variable(value)
        if v.ndim != len(shape):
            value = tensor.shape_padleft(value, len(shape) - v.ndim)
        return ([v] + inputs, outputs)
```
But I didn't like that you couldn't directly see the `make_node` arguments any more (`GpuAlloc` doesn't even define `make_node` then), so I settled for calling the super class `make_node` and pulling out the inputs and outputs of the apply node. Are you fine with the `get_apply_args` route?

By the way, re-reading the code, isn't the `value = ` line totally wrong? Shouldn't it be `v` instead of `value` in both occurrences? I didn't touch that part.